contrib/aws/aws-sdk-go-v2/aws/aws.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+const (
+	tagAWSAgent     = "aws.agent"
+	tagAWSService   = "aws.service"
+	tagAWSOperation = "aws.operation"
+	tagAWSRegion    = "aws.region"
+	tagAWSRequestID = "aws.request_id"
+)
+
+type spanTimestampKey struct{}
+
+// AppendMiddleware takes the aws.Config and adds the Datadog tracing middleware into the APIOptions middleware stack.
+// See https://aws.github.io/aws-sdk-go-v2/docs/middleware for more information.
+func AppendMiddleware(awsCfg *aws.Config, opts ...Option) {
+	cfg := &config{}
+
+	defaults(cfg)
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	tm := traceMiddleware{cfg: cfg}
+	awsCfg.APIOptions = append(awsCfg.APIOptions, tm.initTraceMiddleware, tm.startTraceMiddleware, tm.deserializeTraceMiddleware)
+}
+
+type traceMiddleware struct {
+	cfg *config
+}
+
+func (mw *traceMiddleware) initTraceMiddleware(stack *middleware.Stack) error {
+	return stack.Initialize.Add(middleware.InitializeMiddlewareFunc("InitTraceMiddleware", func(
+		ctx context.Context, in middleware.InitializeInput, next middleware.InitializeHandler,
+	) (
+		out middleware.InitializeOutput, metadata middleware.Metadata, err error,
+	) {
+		// Bind the timestamp to the context so that we can use it when we have enough information to start the trace.
+		ctx = context.WithValue(ctx, spanTimestampKey{}, time.Now())
+		return next.HandleInitialize(ctx, in)
+	}), middleware.Before)
+}
+
+func (mw *traceMiddleware) startTraceMiddleware(stack *middleware.Stack) error {
+	return stack.Initialize.Add(middleware.InitializeMiddlewareFunc("StartTraceMiddleware", func(
+		ctx context.Context, in middleware.InitializeInput, next middleware.InitializeHandler,
+	) (
+		out middleware.InitializeOutput, metadata middleware.Metadata, err error,
+	) {
+		operation := awsmiddleware.GetOperationName(ctx)
+		serviceID := awsmiddleware.GetServiceID(ctx)
+
+		opts := []ddtrace.StartSpanOption{
+			tracer.SpanType(ext.SpanTypeHTTP),
+			tracer.ServiceName(serviceName(mw.cfg, serviceID)),
+			tracer.ResourceName(fmt.Sprintf("%s.%s", serviceID, operation)),
+			tracer.Tag(tagAWSRegion, awsmiddleware.GetRegion(ctx)),
+			tracer.Tag(tagAWSOperation, awsmiddleware.GetOperationName(ctx)),
+			tracer.Tag(tagAWSService, serviceID),
+			tracer.StartTime(ctx.Value(spanTimestampKey{}).(time.Time)),
+		}
+		if !math.IsNaN(mw.cfg.analyticsRate) {
+			opts = append(opts, tracer.Tag(ext.EventSampleRate, mw.cfg.analyticsRate))
+		}
+		span, spanctx := tracer.StartSpanFromContext(ctx, fmt.Sprintf("%s.request", serviceID), opts...)
+
+		// Handle initialize and continue through the middleware chain.
+		out, metadata, err = next.HandleInitialize(spanctx, in)
+		if err != nil {
+			span.Finish(tracer.WithError(err))
+		} else {
+			span.Finish()
+		}
```

Comment:
I think we can just call `span.Finish(tracer.WithError(err))`. If `err` is `nil`, no error will be appended.

Reply:
Yep, you're spot on!